src/wallet/wallet.cpp

```diff
@@ -1147,6 +1153,50 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
 
 
 
+void CWallet::BlockUntilSyncedToCurrentChain() {
+    const CBlockIndex* initialChainTip;
+    {
+        LOCK(cs_main);
+        initialChainTip = chainActive.Tip();
+    }
+    AssertLockNotHeld(cs_main);
+    AssertLockNotHeld(cs_wallet);
+    std::unique_lock<std::mutex> lock(lastBlockProcessedMutex);
+
+    assert(lastBlockProcessed);
+
+    auto pred = [this, initialChainTip] {
+            if (this->lastBlockProcessed == initialChainTip) {
+                return true;
+            }
+            if (this->lastBlockProcessed->GetAncestor(initialChainTip->nHeight) == initialChainTip) {
+                return true;
+            }
+            // Catch the race condition where the wallet may have caught up and
+            // moved past initialChainTip through a reorg before we could get
+            // lastBlockProcessedMutex.
+            // This should be exceedingly rare in regular usage, so potentially
+            // eating 100ms to retry this lock should be fine (not TRY_LOCKing
+            // here would be a lock inversion against lastBlockProcessedMutex)
+            TRY_LOCK(cs_main, mainLocked);
```

Comment:
In commit "Add CWallet::BlockUntilSyncedToCurrentChain()"

Maybe consider dropping the try-lock and replacing it with `lastBlockProcessedMutex.unlock(); LOCK(cs_main);  lastBlockProcessedMutex.lock();`. Maybe this would be a little slower in the average case where this code runs (which is rare to begin with), but it would avoid the 100ms worst case, and make the code simpler because you could also replace the while loop and timeout with a plain `cv.wait(lock, pred)` call.